src/dmd/dtemplate.d

```diff
@@ -6237,6 +6237,16 @@ extern (C++) class TemplateInstance : ScopeDsymbol
             return false;
         }
 
+        // Always emit templates from core.internal.dassert that are instantiated by -checkaction=context
+        // because we might link with code that was compiled with -checkaction=[halt|C|D] / without asserts
+        if (global.params.checkAction == CHECKACTION.context)
+        {
+            assert(this.tempdecl);
+            auto mod = this.tempdecl.getModule();
+            if (mod.hasName(Id.core, Id.internal, Id.dassert))
```

Comment:
Ah now I see what you meant. Yeah, the *exact* module some instance is appended to is defined by https://github.com/dlang/dmd/blob/b6985bb2896477033bd34760dc7b9665c315c576/src/dmd/dtemplate.d#L7372-L7388
and so AFAICT not really predictable and highly depending on the command-line (`importedFrom`).

I'd check the identifier of `this.tempdecl` instead, and then checking its declaring module - as codegen'ing it should lead to all children instances getting codegen too. As mentioned earlier, I'd try the allinst branch in that case.

Note that this `needsCodegen()` function is likely performance-sensitive and called many times (e.g., when traversing the `tinst` parents of children); setting up `minst` has been enough as cache for now (tinst/tnext chains are disconnected after the 1st invocation).